remme/ws/events.py

```diff
@@ -0,0 +1,168 @@
+    entities_changed = None
+    data = {}
+    for item in attributes:
+        if item.key == 'header_signature':
+            transaction_id = item.value
+        if item.key == 'entities_changed':
+            entities_changed = json.loads(item.value)
+
+    if type.startswith(ATOMIC_SWAP):
+        for entity in entities_changed:
+            if entity['type'] == AtomicSwapInfo.__name__:
+                data = entity
+    if not data:
+        data = entities_changed
+    return transaction_id, data
+
+
+class WSEventSocketHandler(BasicWebSocketHandler):
+    def __init__(self, stream, loop):
+        super().__init__(stream, loop)
+        # events to subscribers
+        self._events = {event.value: [] for event in Events}
+        self._events_updator_task = weakref.ref(
+            asyncio.ensure_future(
+                self.listen_events(), loop=self._loop))
+        self.subscribe_events()
+
+    # return what value to be mapped to web_sock
+    async def subscribe(self, web_sock, entity, data):
+        if entity == Entity.EVENTS:
+            events = data.get('events', [])
+
+            if not events:
+                raise SocketException(web_sock, Status.EVENTS_NOT_PROVIDED, f"Events being subscribed are not provided")
+
+            for event in events:
+                if event not in self._events:
+                    raise SocketException(web_sock, Status.WRONG_EVENT_TYPE, f"Event: {event} is not supported")
+                if web_sock in self._events[event]:
+                    raise SocketException(web_sock, Status.ALREADY_SUBSCRIBED, f"Socket is already subscribed to: {event}")
+                self._events[event] += [web_sock]
+
+            LOGGER.info(f'Events being subscribed to: {events}')
+
+            return {'events': events}
+
+    def unsubscribe(self, web_sock):
+        # currently web_sock notification is based on self._subscriptions list
+        for event, web_socks in self._events.items():
+            if web_sock in web_socks:
+                web_socks.remove(web_sock)
+                self._events[event] = web_socks
+
+        pass
+
+    def subscribe_events(self):
+        # Setup a connection to the validator
+        LOGGER.info(f"Subscribing to events")
+        ctx = zmq.Context()
+        self._socket = ctx.socket(zmq.DEALER)
+        self._socket.connect(ZMQ_URL)
+        LOGGER.info(f"Connected to ZMQ")
+
+        request = ClientEventsSubscribeRequest(subscriptions=self._make_subscriptions(), last_known_block_ids=[]).SerializeToString()
+
+        # Construct the message wrapper
+        correlation_id = generate_random_key()  # This must be unique for all in-process requests
+        msg = Message(
+            correlation_id=correlation_id,
+            message_type=Message.CLIENT_EVENTS_SUBSCRIBE_REQUEST,
+            content=request)
+
+        # Send the request
+        LOGGER.info(f"Sending subscription request.")
+        self._socket.send_multipart([msg.SerializeToString()])
```

Comment:
Required to be async

Reply:
Example taken from here:
https://sawtooth.hyperledger.org/docs/core/releases/1.0/app_developers_guide/event_subscriptions.html#submitting-an-event-subscription

Reply:
It must be async in the event loop for preventing of holding a whole loop. Fast way to do - https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.run_in_executor

Reply:
<img width="682" alt="screen shot 2018-08-15 at 3 48 57 pm" src="https://user-images.githubusercontent.com/6625496/44148920-d6a98dce-a0a2-11e8-98e8-4ff0394443cf.png">

Reply:
Sync call to the socket and do not release a loop.

Reply:
subscribe_events is ok to be synced. it is called one time during start-up